### DISCUSSION TITLE
Why is `data` in `onSuccess` of `useAction` optional?

### DISCUSSION THREAD

**Discussion by maltesa** [next-safe-action/next-safe-action#214] [Q&A] [answered]
Hi,

I was wondering why `data` is optional in the following example:

```typescript
useServerAction(createInstituteAction, {
  onSuccess: ({ data }) => {
    // Expecting data to have the type of `createInstituteAction`s return value
    // but value is: <createInstituteAction return value> | undefined
  }
})
```

Since the actions succeed, the value must be set, right?

**Comment by TheEdoRan** [comment, marked as answer]
See:
- #175 

The reason is that the action runs `onSuccess` callback even when you use `redirect`, which internally throws an error caught by Next.js to perform the redirect. In this case, `data` is undefined.

**Reply by maltesa** [reply]
Ah great. I searched for this in issues and discussions but didn't find it. Sorry 👍🏻
